### PR TITLE
Rich is here!

### PR DESCRIPTION
**Pull request by puzzlepeaches**
Good morning!

In this pull request, the following is accomplished:

* All text output displayed to the end user is now done using the rich library.
* All tables output to the end user uses the rich live table functionality. 
* All code has been reformatted using the black library. Partially using the article [here](https://ljvmiranda921.github.io/notebook/2018/06/21/precommits-using-black-and-flake8/) on my box only to format on git commit. 
* Pulled in changes made in master as well as I could.

Some additional notes:

* Due to the large amount of refactoring that had to occur here, we should for sure do some additional testing to make sure everything is square before merging.
* I was able to confirm the OWA, NTLM, and O365 modules locally. All others will need to be tested to ensure reliability. 
* Theme set for rich is not set in stone. Feel free to make changes based on your preference.

Coming soon to a theater near you:

* Basic proxy option to support custom Fireprox setup a little more easily.
* Backoff option for O365 module to deal with smart-lockout and potentially an additional module based on [this](https://github.com/MarkoH17/Spray365) tool. (should chat on this one)
* Refactoring of output CSV file name to be unique as opposed to overwriting on each run.
* A resume option to start a spray again if something stopped it abruptly. 
* Live highlighting in tabled output to call out anomalous login attempts. (should chat on this one)

### CODE REVIEW

**Comment by Tw1sm** [comment]
I'll work on testing this week - looks like most of the changes are visual so I imagine it'll be easy. I'm thinking I'll refactor the main spraycharles.py soon to utilize a class, then we can cut down on all the args getting passed around.

File: spraycharles.py [hunk 1]
```diff
@@ -13,192 +13,517 @@
...
     except (requests.ConnectionError, requests.ReadTimeout) as e:
-        colors.color_print('\n[!] Connection error - sleeping for 5 seconds', colors.red)
+        console.print("\n[!] Connection error - sleeping for 5 seconds", style="danger")
         sleep(5)
-        login(target, username, password, csvfile)
+        login(target, username, password, csvfile, table)
+
+    if name.lower() == "office365":
+        return result, message, ruser, rpass, rcode, rlength
+    else:
```
Comment (Tw1sm) [inline spraycharles.py:288]: SMB needs handled here

```suggestion
    elif name.lower() == "smb":
        return ruser, rpass, rcode
     else:
```

**Review by Tw1sm — CHANGES_REQUESTED** [review]
Alright, got around to reviewing a little more in-depth after work, although, I did need to password spray on an engagement today so I did confirm the Office365 module and output work. Rich library for output/tables gives it a very smooth, modern feel, I like the switch!

I left some review comments mostly around debugging crashes I encountered with testing the SMB module. I think those changes will be easy to make

Biggest thing I noticed that needs to be addressed, is when I was spraying 365 today with a big list, the live table only fills up the terminal window, then just displays `...` at the bottom and doesn't show the results after or let you scroll down to see them. It seemed to render everything and let you scroll through when the table was complete, either at the end of the spray (maybe spray interval) or ctrl-c. I think this has to do with what Rich calls [vertical overflow](https://rich.readthedocs.io/en/stable/live.html?highlight=vertical%20overflow#vertical-overflow).  I tried setting `vertical_overflow="visible"` and testing, but I still got the ellipsis occasionally. Was seeing this behavior in a screen session in a funky ZSH shell and in/out of screen in the standard Kali ZSH shell when I tested tonight. Let me know if you've got any ideas here, I'd love for it to keep scrolling down as more entries to the table get added, similar to how it works in master.

**Comment by puzzlepeaches** [comment]
Looked over your notes and attempted to fix everything mentioned. I opted to sunset the OpenVPN module. It was a one time use thing and I can't confirm it will consistently work without setting up an internal instance. Therefore, I think we should just pull it from the repo for the time being. Was going to mention that I thought we should do this soon regardless. Thoughts?

As for what you mentioned surrounding live table rendering - I don't have any specific ideas on how to fix the ellipses/rendering issues. That totally slipped through the cracks as it wasn't showing up during my testing. I'm going to pull this fork in an an op I'm working currently with a much larger list of users to hopefully identify and implement a fix. Will keep you updated and let you know what I find out.

**Comment by Tw1sm** [comment]
Sunsetting the OpenVPN module is fine with me - personally haven't seen OpenVPN on an engagement and we can bring it back from an old commit if needed.

That reminds me, based on the prior PR, was your plan to remove the `--pause` flag after finding a fix for repeat notifications, or did you intend to keep it regardless? (Kind of unrelated to this PR)

**Comment by puzzlepeaches** [comment]
> That reminds me, based on the prior PR, was your plan to remove the `--pause` flag after finding a fix for repeat notifications, or did you intend to keep it regardless? (Kind of unrelated to this PR)

I would like to keep it included if that's okay with you. I can definitely see myself using it on certain projects.

**Comment by Tw1sm** [comment]
Keeping it is cool with me

**Comment by Tw1sm** [comment]
Closing PR as this was superseded by #11. Rich library live table proved to be a bit too unstable for the use case here

**Pull request closed without merging**